File: 2022/scala/src/main/scala/Day09.scala
enum Direction:
  case Left, Right, Up, Down

object Day09 {
  def solve = {
    val input = Util.readResource("Day09/input.txt")
    println(part1(input))
    println(part2(input))
  }

  def part1(input : List[String]) : Int = {
    val commands = parseCommands(input)
    applyCommands(commands,  List.fill(2)((0, 0)))
      .map(_.last)
      .distinct
      .size
  }

  def part2(input : List[String]) : Int = {
    val commands = parseCommands(input)
    applyCommands(commands,  List.fill(10)((0, 0)))
      .map(_.last)
      .distinct
      .size
  }

  def applyCommands(commands : List[Direction], positions : List[(Int, Int)]) : List[List[(Int, Int)]] = {
    commands.foldLeft(List(positions)) { (movements, direction) =>
      val currentPosition = movements.head
      val head = currentPosition.head
      val newHeadPosition = move(direction, head)
      val newPositions = currentPosition.tail.foldLeft(List(newHeadPosition))((rope, curr) => {
        rope :+ moveKnot(rope.last, curr)
      })
      newPositions :: movements
    }
  }

  def moveKnot(head : (Int, Int), knot : (Int, Int)) : (Int, Int) = {
    val xDis = (head._1 - knot._1).abs
    val yDis = (head._2 - knot._2).abs
    // not far enough away to move the knot
    // Chebyshev distance
    if (xDis.max(yDis) <= 1) {
      return knot
    }

    val newX = compareAndMove(knot._1, head._1)
    val newY = compareAndMove(knot._2, head._2)
    (newX, newY)
  }

  def compareAndMove(a : Int, b : Int) : Int = {
    if (a > b) {
      a - 1
    } else if (a < b) {
      a + 1
    } else {
      a
    }
  }

  def move(direction : Direction, point : (Int, Int)) : (Int, Int) = {
    direction match {
      case Direction.Left => (point._1 - 1, point._2)
      case Direction.Right => (point._1 + 1, point._2)
      case Direction.Up => (point._1, point._2 + 1)
      case Direction.Down => (point._1, point._2 - 1)
    }
  }

  def parseCommands(input : List[String]) : List[Direction] = {
    input.flatMap(line => {
      val (direction, amount) = line.splitAt(1)
      1.to(amount.trim.toInt).map(idx => {
        direction.trim match {
          case "R" => Direction.Right
          case "U" => Direction.Up
          case "L" => Direction.Left
          case "D" => Direction.Down
          case _ => throw new Exception(s"what? $direction $amount")
        }
      })
    })
  }
}

File: 2022/rust/src/bin/day09.rs
use std::{collections::HashSet, fs::read_to_string};

#[derive(Clone, Copy, Debug)]
enum Direction {
    Up,
    Down,
    Left,
    Right,
}

#[derive(Debug)]
struct HeadKnot {
    position: (i32, i32),
    tail: Vec<TailKnot>,
}

#[derive(Debug)]
struct TailKnot {
    position: (i32, i32),
    positions_visited: HashSet<(i32, i32)>,
}

impl HeadKnot {
    fn move_direction(&mut self, direction: &Direction) {
        match direction {
            Direction::Up => self.position.1 += 1,
            Direction::Down => self.position.1 -= 1,
            Direction::Left => self.position.0 -= 1,
            Direction::Right => self.position.0 += 1,
        }
        let mut curr_pos = self.position;
        for knot in self.tail.iter_mut() {
            knot.maybe_move(&curr_pos);
            curr_pos = knot.position;
        }
    }
}

impl TailKnot {
    fn maybe_move(&mut self, lead_position: &(i32, i32)) {
        self.positions_visited.insert(self.position);
        // only move if lead is 2 or more away
        if self.distance_from(lead_position) < 2 {
            return;
        }

        if self.position.0 > lead_position.0 {
            self.position.0 -= 1;
        } else if self.position.0 < lead_position.0 {
            self.position.0 += 1;
        }

        if self.position.1 > lead_position.1 {
            self.position.1 -= 1;
        } else if self.position.1 < lead_position.1 {
            self.position.1 += 1;
        }
        self.positions_visited.insert(self.position);
    }

    fn distance_from(&self, lead_position: &(i32, i32)) -> u32 {
        let x_distance = lead_position.0.abs_diff(self.position.0);
        let y_distance = lead_position.1.abs_diff(self.position.1);
        std::cmp::max(x_distance, y_distance)
    }
}

fn main() {
    let input = read_to_string("src/inputs/09.txt").unwrap();
    part1(&input);
    part2(&input);
}

fn part1(input: &String) {
    let directions = parse_input(input);
    let mut head = create_rope(2);
    for direction in directions.iter() {
        head.move_direction(direction);
    }

    let count = head.tail.last().unwrap().positions_visited.len();
    println!("{}", count);
}

fn part2(input: &String) {
    let directions = parse_input(input);
    let mut head = create_rope(10);
    for direction in directions.iter() {
        head.move_direction(direction);
    }

    let count = head.tail.last().unwrap().positions_visited.len();
    println!("{}", count);
}

fn create_rope(knots: u32) -> HeadKnot {
    let mut tail = vec![];
    for _ in 0..(knots - 1) {
        tail.push(TailKnot {
            position: (0, 0),
            positions_visited: HashSet::new(),
        });
    }

    HeadKnot {
        position: (0, 0),
        tail,
    }
}

fn parse_input(input: &String) -> Vec<Direction> {
    input
        .lines()
        .map(|line| line.split_once(" ").unwrap())
        .flat_map(|(raw_dir, times)| {
            let direction = match raw_dir {
                "R" => Direction::Right,
                "L" => Direction::Left,
                "U" => Direction::Up,
                "D" => Direction::Down,
                _ => panic!("woops"),
            };
            let times: usize = times.parse().unwrap();
            vec![direction; times]
        })
        .collect::<Vec<Direction>>()
}
